File: NapakalakiJava/src/NapakalakiGame/NumericBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author danidiaz
 */
public class NumericBadConsequence extends BadConsequence {
    private int nVisibleTreasures;
    private int nHiddenTreasures;

   /** Cartas que te hacen perder niveles y objetos.
     * @param text Texto de la consecuencia
     * @param levels Niveles perdidos
     * @param nVisible Número de objetos visibles que pierdes
     * @param nHidden Número de objetos ocultos que pierdes
     */
    public NumericBadConsequence(String text, int levels, int nVisible, int nHidden) // NBC
    {
        super(text,levels);
        nVisibleTreasures = nVisible;
        nHiddenTreasures = nHidden;
    }
    
    public int getnVisibleTreasures(){return nVisibleTreasures;}
    public int getnHiddenTreasures(){return nHiddenTreasures;} 

    @Override
    public boolean isEmpty() {
        return (nHiddenTreasures == 0 && nVisibleTreasures == 0);
    }

    @Override
    public NumericBadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h) {
        NumericBadConsequence new_bc = this; // Si no hay que hacer nada, la nueva bc es ella misma
        
        if (nVisibleTreasures > 0 || nHiddenTreasures > 0)
        {
            // Si el número de tesoros del jugador es menor que
            // los tesoros que tenemos que perder según la bc,
            // actualizamos su valor
            int new_nVisibleTreasures = 0, new_nHiddenTreasures = 0;
            
            if (v.size() < nVisibleTreasures)
                new_nVisibleTreasures = v.size();

            if (h.size() < nHiddenTreasures)
                new_nHiddenTreasures = h.size();
            
            new_bc = new NumericBadConsequence(getText(), getLevels(),
                        new_nVisibleTreasures,
                        new_nHiddenTreasures
                        );
        }
        
        return new_bc;
    }

    @Override
    public void substractVisibleTreasure(Treasure t) {
        if (nVisibleTreasures > 0)
            nVisibleTreasures--;
    }

    @Override
    public void substractHiddenTreasure(Treasure t) {
        if (nHiddenTreasures > 0)
            nHiddenTreasures--;
    }
    
    @Override
    public String toString(){
        return super.toString();
    }
}

File: NapakalakiRuby/lib/numeric_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'bad_consequence'

module NapakalakiGame
  class NumericBadConsequence < BadConsequence
    attr_reader :nVisibleTreasures, :nHiddenTreasures
    def initialize(text, levels, someVisibleTreasures, someHiddenTreasures)
      super(text,levels)
      @nVisibleTreasures=someVisibleTreasures
      @nHiddenTreasures=someHiddenTreasures
    end
    public_class_method :new

    def isEmpty
      return (@nHiddenTreasures == 0 && @nVisibleTreasures == 0)
    end

    def adjustToFitTreasureLists(v,h)

        new_bc = self # Si no hay que hacer nada, la nueva bc es ella misma

        if (@nVisibleTreasures > 0 || @nHiddenTreasures > 0)

          new_nVisibleTreasures = 0
          new_nHiddenTreasures = 0

          if (v.size < @nVisibleTreasures)
            new_nVisibleTreasures = v.size
          end

          if (h.size < @nHiddenTreasures)
            new_nHiddenTreasures = h.size
          end

          new_bc = NumericBadConsequence.new(@text,@levels,
                  new_nVisibleTreasures, new_nHiddenTreasures)
        end

        return new_bc

      end


      def substractVisibleTreasure(t)
        if (@nVisibleTreasures > 0)
            @nVisibleTreasures = @nVisibleTreasures - 1
        end
      end


      def substractHiddenTreasure(t)
        if (@nHiddenTreasures > 0)
            @nHiddenTreasures = @nHiddenTreasures - 1
        end
      end

      def to_s
        return "text: #{@text} 
        levels: #{@levels}
        nVisibleTreasures: #{@nVisibleTreasures}
        nHiddenTreasures: #{@nHiddenTreasures}
        "
      end
  end
end